lib/chibi/apropos.scm

```diff
@@ -0,0 +1,36 @@
+(define (list-bindings env)
+  (let parents ((env env) (binds '()))
+    (if (not env) binds
+        (let symbols ((syms (env-exports env)) (binds binds))
+          (if (null? syms) (parents (env-parent env) binds)
+              (symbols (cdr syms) (if (assv (car syms) binds) binds
+                                      (cons (cons (car syms) env)
+                                            binds))))))))
+
+(define (apropos-list-bindings query)
+  (cond ((symbol? query) (set! query (symbol->string query)))
+        ((not (string? query))
+         (error "Apropos query must be a symbol or a string")))
+  (sort (filter (lambda (binding)
+                  (string-contains (symbol->string (car binding)) query))
+                (list-bindings (interaction-environment)))
+        (lambda (a b) (string<? (symbol->string (car a))
+                                (symbol->string (car b))))))
+
+(define (apropos-list query) (map car (apropos-list-bindings query)))
+
+(define (apropos-prefix sym env)
+  (let ((p "procedure  ")
+        (s "syntax     ")
+        (v "variable   "))
+    (call/cc (lambda (bail)
```

Comment:
No need for call/cc, you can just return from the exception handler normally.

Reply:
How would you write that? `(with-exception-handler (lambda (_) s) (lambda () (eval sym env)))` causes `ERROR: exception handler returned` when `sym` is a macro

Reply:
Oh, you would need guard, if you want the lighter with-exception-handler you do need call/cc.

Reply:
OK, right. I force-pushed a new commit changing it to use `guard` instead. It's still a bit of a hack in the absence of an `env-ref` in Scheme - a full `eval` is not really needed as all bindings should point to self-evaluating objects.